src/cljs/athens/events/remote.cljs

```diff
@@ -84,79 +85,137 @@
-      (doseq [[id event] in-memory-events]
-        (log/debug ":remote/update-optimistic-state reapply optimistic event id" (pr-str id))
-        (swap! reapplied-tx-data assoc
-               (utils/uuid->string id)
-               (atomic-resolver/resolve-transact! db/dsdb event)))
-
-      ;; Update rollback-tx-data with the new tx-data.
-      (update db :remote/rollback-tx-data merge @reapplied-tx-data))))
+(defn- promotion?
+  [[type _ _ _ noop?]]
+  (and (= type :promote) (not noop?)))
+
+
+(defn- rollback!
+  "Revert all events in the :memory stage of event-sync via undo.
+  Should only be used followed by rollforward."
+  [[db conn]]
+  (let [in-memory-events       (-> db :remote/event-sync :stages :memory)
+        rollback-tx-data       (-> db :remote/rollback-tx-data)
+        count-in-memory-events (count in-memory-events)
+        rolled-back-events     (atom #{})]
+
+    (if (= count-in-memory-events 0)
+      (log/debug "rollback skipped, nothing to rollback")
+      (do
+        (log/debug "rollback" count-in-memory-events "events")
+        ;; Undo events in the reverse order (e.g. first undo reverse last event).
+        (doseq [[id event]  (reverse in-memory-events)]
+          (let [id          (utils/uuid->string id)
+                tx          (rollback-tx-data id)
+                rollback-tx (undo-tx-data tx)]
+            (if (nil? tx)
+              (do
+                ;; It's very bad if this happens. It means we're not rolling back one of the events.
+                ;; It shouldn't ever happen, but we should keep an eye out for it in the beta console logs.
+                (log/warn "rollback no tx-data found for" id)
+                (log/warn "rollback in-memory-events ids" (keys in-memory-events))
+                (log/warn "rollback rollback-tx-data ids" (keys rollback-tx-data)))
+              (do
+                (log/debug "rollback event id" (pr-str id))
+                (log/debug "rollback event:")
+                (log/debug (with-out-str (pp/pprint event)))
+                (log/debug "rollback original tx:")
+                (log/debug (with-out-str (pp/pprint tx)))
+                (log/debug "rollback rollback tx:")
+                (log/debug (with-out-str (pp/pprint rollback-tx)))
+                (d/transact! conn rollback-tx)
+                (swap! rolled-back-events conj id)))))))
+
+    ;; Remove rolled back txs from rollback-tx-data.
+    (let [db' (update db :remote/rollback-tx-data #(apply dissoc % @rolled-back-events))]
+      [db' conn])))
+
+
+(defn- rollforward!
+  "Apply all events in the :memory stage of event-sync via resolve.
+  Should only be used preceded by rollback."
+  [[db conn]]
+  (let [in-memory-events       (-> db :remote/event-sync :stages :memory)
+        count-in-memory-events (count in-memory-events)
+        reapplied-tx-data      (atom {})]
+
+    (if (= count-in-memory-events 0)
+      (do
+        (log/debug "rollforward skipped, nothing to rollforward")
```

Comment:
It looks like it's unexpected, we've rolledback to rollfarward (rebase), but there is nothing to rebase, correct?

If so bump this to `log/warn`.

Reply:
This can happen on `:remote/reject-forwarded-event`, if the rejected event was the only one or not found, and during `:remote/apply-forwarded-event` if there was nothing rolled back at all. I think those are still legitimate cases.